### ISSUE TITLE
Can't use device queue as argument.

### ISSUE THREAD

**Issue by OmegaDoom** [inducer/pyopencl#246]
Hello. I'm not sure if device side queues are suppored already but I have an error with such code.

```
properties = cl.command_queue_properties.OUT_OF_ORDER_EXEC_MODE_ENABLE or\
             cl.command_queue_properties.ON_DEVICE
device_queue = cl.CommandQueue(context, device, properties)
kernel.set_args(device_queue);
```
error
```
pyopencl.cffi_cl.LogicError: when processing argument #1 (1-based):
a bytes-like object is required, not 'CommandQueue'
```

**Comment by inducer** [comment]
Could you see if this branch does the job for you and report back?

https://gitlab.tiker.net/inducer/pyopencl/merge_requests/56

I don't think I have an ICD available to test with that actually allows
passing device queues.

**Comment by OmegaDoom** [comment]
Thank you. I have checked it and looks like it works. Even default queue works although prebuild module can't work with it too.
But it works only for gpu device. cpu enqueue_nd_range_kernel throws exception - 
"Intel::OpenCL::DeviceBackend::Exceptions::DeviceBackendExceptionBase".
 Equivalent c++ program on same device doesn't have such problem.

**Comment by OmegaDoom** [comment]
It seems to me i have found why it throws exception. It doesn't throw for new kernel source. Andreas do you  use caching? I want to try same for c++ program. I see - clCreateProgramWithBinary
Even gpu crashes with INVALID_BINARY.

**Comment by OmegaDoom** [comment]
It's intel bugs. I tested it on c++ program and the behavior is the same. If there is "enqueue_kernel" in kernel source then clCreateProgramWithBinary, clProgramBuild, clEnqueueNDRange for cpu device throws segfault. Maybe it is any CL2.0+ feature but I tested only "enqueue_kernel".
And if there is "-g" option then cpu segfaults again and gpu clCreateProgramWithBinary  returns INVALID_BINARY. Not sure if the last one really the bug but the first one definitely looks like.  I will report it to intel. What about an options to disable caching? It would be interesting to test on amd but i don't have it.
